### PR TITLE
[Hotfix] Fix fetchLatestNonOptimizedSnapshotTime

### PR DESCRIPTION
**Pull request by XBaith**
## Why are the changes needed?

The condition of fetching the latest non-optimized snapshot is incorrect and the timestamp of expiration is incorrect that must be fixed right away

## Brief change log

-

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (no)
- If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Review by wangtaohz — APPROVED** [review]
LGTM

**Comment by XBaith** [comment]
cc @zhoujinsong

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java [hunk 1]
```diff
@@ -213,22 +213,31 @@ public void expireData(TableRuntime tableRuntime) {
...
+    if (expirationConfig.getBaseOnRule() == DataExpirationConfig.BaseOnRule.CURRENT_TIME) {
+      return Instant.now().atZone(getDefaultZoneId(field)).toInstant();
+    } else {
+      Snapshot snapshot = IcebergTableUtil.getSnapshot(getTable(), false);
+      long currentSnapshotTs =
+          Optional.ofNullable(snapshot).isPresent()
+              ? snapshot.timestampMillis()
+              : System.currentTimeMillis();
+      // snapshot timestamp should be UTC
+      return Instant.ofEpochMilli(
```
Comment (shidayang) [inline ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java:234]: If there is only an "optimizing" snapshot, will the date of this snapshot expire?

Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java:234]: Yes. If there is no non-optimized snapshot, the TableMaintainer will chose the timestamp of  currentSnapshot as the expire time

Comment (shidayang) [reply ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java:234]: If the rule is LAST_COMMIT_TIME and non-optimized snapshots do not exist, should file expiration not be enforced?

Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java:234]: The purpose of keeping a snapshot of the last commit is to preserve historical data, so you're right!

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java [hunk 2]
```diff
@@ -216,22 +216,33 @@ public void expireData(TableRuntime tableRuntime) {
...
+        long lastCommitTimestamp = fetchLatestNonOptimizedSnapshotTime(getTable());
+        // if the table does not exist any non-optimized snapshots, should skip the expiration
+        if (lastCommitTimestamp != Long.MAX_VALUE) {
+          // snapshot timestamp should be UTC
+          return Instant.ofEpochMilli(lastCommitTimestamp).atZone(ZoneOffset.UTC).toInstant();
+        } else {
+          return Instant.MIN;
+        }
+      default:
+        return Instant.MIN;
```
Comment (shidayang) [inline ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java:242]: Why is MIN? Maybe throw exception and catch it in outside?

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java [hunk 3]
```diff
@@ -216,22 +216,33 @@ public void expireData(TableRuntime tableRuntime) {
...
+    switch (expirationConfig.getBaseOnRule()) {
+      case CURRENT_TIME:
+        return Instant.now().atZone(getDefaultZoneId(field)).toInstant();
+      case LAST_COMMIT_TIME:
+        Snapshot snapshot = IcebergTableUtil.getSnapshot(getTable(), false);
+        long lastCommitTimestamp = fetchLatestNonOptimizedSnapshotTime(getTable());
+        // if the table does not exist any non-optimized snapshots, should skip the expiration
+        if (lastCommitTimestamp != Long.MAX_VALUE) {
+          // snapshot timestamp should be UTC
+          return Instant.ofEpochMilli(lastCommitTimestamp).atZone(ZoneOffset.UTC).toInstant();
```
Comment (shidayang) [inline ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java:237]: Is atZone unnecessary?

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java (hunk 2 above)
Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java]: sure, will throw exception by default

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java (hunk 3 above)
Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java]: removed

**Review by shidayang — APPROVED** [review]
LGTM

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java [hunk 4]
```diff
@@ -20,6 +20,11 @@
 
 import static org.apache.iceberg.relocated.com.google.common.primitives.Longs.min;
 
+import com.google.common.annotations.VisibleForTesting;
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java:23]: Maybe we should depend on the Iceberg shaded classes.

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/MixedTableMaintainer.java [hunk 5]
```diff
@@ -21,6 +21,9 @@
 import static com.netease.arctic.utils.ArcticTableUtil.BLOB_TYPE_OPTIMIZED_SEQUENCE_EXIST;
 import static org.apache.iceberg.relocated.com.google.common.primitives.Longs.min;
 
+import com.google.common.annotations.VisibleForTesting;
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/MixedTableMaintainer.java:24]: Maybe we should depend on the Iceberg shaded classes.

File: ams/server/src/main/java/com/netease/arctic/server/utils/IcebergTableUtil.java [hunk 6]
```diff
@@ -18,6 +18,9 @@
 
 package com.netease.arctic.server.utils;
 
+import com.google.common.base.Predicate;
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/utils/IcebergTableUtil.java:21]: Maybe we should depend on the Iceberg shaded classes.

File: ams/server/src/main/java/com/netease/arctic/server/utils/IcebergTableUtil.java [hunk 7]
```diff
@@ -78,9 +80,10 @@ public static Snapshot getSnapshot(Table table, boolean refresh) {
     return table.currentSnapshot();
   }
 
-  public static Optional<Snapshot> findSnapshot(Table table, Predicate<Snapshot> predicate) {
-    Iterable<Snapshot> snapshots = table.snapshots();
-    return Iterables.tryFind(snapshots, predicate);
+  public static Optional<Snapshot> findSnapshotDesc(Table table, Predicate<Snapshot> predicate) {
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/utils/IcebergTableUtil.java:83]: I am not sure `findSnapshotDesc` is a good enough name for this method.
IMO, this method seems to try to find a snapshot matching a specific condition, so maybe `findFirstMatchSnapshot` is a proper name?

**Review by zhoujinsong — COMMENTED** [review]
@XBaith I left some comments, please take a look.

**Review by zhoujinsong — APPROVED** [review]
LGTM.

**Pull request merged**